### PR TITLE
 Router Helpers & Modifiers

### PR DESCRIPTION
**Pull request by chadhietala**

### CODE REVIEW

File: text/0000-router-helpers.md [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+```
+
+In the DOM you will have an `href` on the anchor that gets serializes as:
+
+```html
+<a href="/profile?someBool=true" class="active ember-view">Profile</a>
+```
+
+Looking at a template you would have no idea that rendering the `{{link-to}}` would result in the query params being serialized. From an implementation point of view, this is problematic as we are forced to `lookup` the `Route` and the associated `Controller` to grab the query params. This can add a non-trival amount of overhead during rendering, especially if you have many `{{link-to}}`s on a route that link many different parts of your application. As a sidenote, this is one of the things  that needs to be delt with if we are ever to kill controllers.
+
```
Comment (chadhietala) [inline text/0000-router-helpers.md:80]: TODO:

Add Angle Bracket Invocation + Positional Params Issues

File: text/0000-router-helpers.md [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+# Router Helpers & Modifiers
+
+## Summary
+
+This RFC introduces new router helpers that represent a decomposition of functionality of what we commonly use `{{link-to}}` for. This RFC also deprecates the `{{link-to}}` component, `{{query-params}}` helper, and default query param serialization. Below is a list of the new helpers:
+
+```hbs
+{{transition-to}}
+{{url-for}}
```
Comment (chancancode) [inline text/0000-router-helpers.md:13]: url-for vs path-for?

File: text/0000-router-helpers.md [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+
+### Does Not Work With Angle Bracket Invocation
+
+Since angle bracket invocation does not support positional params, `{{link-to}}` has to adapt it's public API.
+
+## Detailed design
+
+Below is a detailed design of all of the template helpers and modifiers.
+
+### `{{transition-to}}` Element Modifier
```
Comment (chancancode) [inline text/0000-router-helpers.md:85]: probably should have a subexpression form also

**Comment by ef4** [comment]
Please make it clear that states like `is-active` are _context dependent_, not global.

In particular, you might have two copies of `(is-active "posts")` in your app simultaneously where one is true and one is false, because you're in the middle of an animated transition, or because you're pre-rendering a route that hasn't been entered yet. These scenarios are already implied by the features in [RFC 95](https://github.com/emberjs/rfcs/blob/master/text/0095-router-service.md).

In implementor terms (which probably don't go directly into the RFC text, but I know you will understand): these helpers should consume the current RouteInfo, as opposed to binding directly to the router service.

I think all of them should be reasonable to implement in terms of RouteInfo, except perhaps `is-transitioning-{in,out}`, which could probably use a combination of RouteInfo and the router service to identify how their particular context fits into the current Transition.

Is there any support in here for constructing route-relative URLs? If so, the same caveat applies to them: they must be relative to the current context. This is a thing that breaks painfully during animations today with `link-to`, see https://github.com/ember-animation/liquid-fire/issues/347

**Comment by chadhietala** [comment]
@ef4 Yes this RFC should force the completion of [RFC 95](https://github.com/emberjs/rfcs/blob/master/text/0095-router-service.md)

File: text/0000-router-helpers.md (hunk 2 above)
Comment (chadhietala) [reply text/0000-router-helpers.md:13]: These are more less intended declarative APIs that map directly on to the router service methods. Changing the name would introduce some incoherence here.

**Comment by chadhietala** [comment]
Opened up against RFC repo

**Pull request closed without merging**